crates/env/src/topics.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Docs
+
+use crate::EnvTypes;
+
+/// The concrete implementation that is guided by the topics builder.
+///
+/// To be implemented by the on-chain and off-chain environments respectively.
+#[doc(hidden)]
+pub trait TopicsBuilderBackend<E>
+where
+    E: EnvTypes,
+{
+    /// The type of the serialized event topics.
+    type Output;
+
+    /// Initialized the backend with the expected number of event topics.
+    fn expect(&mut self, expected_topics: usize);
+
+    /// Pushes another topic for serialization to the backend.
+    fn push_topic<T>(&mut self, topic_value: &T)
+    where
+        T: scale::Encode;
+
+    /// Extracts the serialized topics.
+    fn output(self) -> Self::Output;
+}
+
+/// Builder for event topic serialization.
+///
+/// Abstraction to build up event topic serialization with zero-overhead,
+/// no heap-memory allocations and no dynamic dispatch.
+#[doc(hidden)]
+pub struct TopicsBuilder<S, E, B> {
+    backend: B,
+    state: core::marker::PhantomData<fn() -> (S, E)>,
+}
+
+impl<E, B> From<B> for TopicsBuilder<state::Uninit, E, B>
+where
+    E: EnvTypes,
+    B: TopicsBuilderBackend<E>,
+{
+    fn from(backend: B) -> Self {
+        Self {
+            backend,
+            state: Default::default(),
+        }
+    }
+}
+
+#[doc(hidden)]
+pub mod state {
+    /// The topic builder is uninitialized and needs to be provided with the
+    /// expected number of topics that need to be constructed.
+    pub enum Uninit {}
+    /// There are some remaining topics that need to be provided with some values.
+    pub enum HasRemainingTopics {}
+    /// There are no more remaining topics and the topic builder shall be finalized.
+    pub enum NoRemainingTopics {}
+}
+
+impl<E, B> TopicsBuilder<state::Uninit, E, B>
+where
+    E: EnvTypes,
+    B: TopicsBuilderBackend<E>,
+{
+    /// Initializes the topics builder and informs it about how many topics it must expect to serialize.
+    ///
+    /// The number of expected topics is given implicitely by the `E` type parameter.
+    pub fn build<Event: Topics>(
+        mut self,
+    ) -> TopicsBuilder<<Event as Topics>::RemainingTopics, E, B> {
```

Comment:
I don't get how this works. `<Event as Topics>::RemainingTopics` does not implement `SomeRemainingTopics`.

Reply:
It does not need to. That is only going to be important for the `push_topics` method.
With this abstraction we simply can make sure that we do not specify more event topics than there actually are. However, we do all this on a purely type level.

Reply:
But how is a `TopicsBuilder` to which the impl block with `push_topics` applies created? Thought it should be created through `build`.

Reply:
Not sure I get your question but the on/off chain environment first initialize the `TopicsBuilder` in an uninitialized state. Then, through `Topics` trait it gets first initialized with the amount of expected topics and with every call to `push_topic` a type-based counter is decreased so that exactly `N` topics are going to be serialized before the implementer of `Topics` can finally call `finish` to yield the result back to the environment for doing the rest of the serialization efforts.

Reply:
Note that this type state approach was taken as an extra safety check between environment implementation and ink! codegen and to also eliminate some assertions that would have been necessary otherwise.

Reply:
Thanks for the explainer. I got it now.